esp-hal/src/rtc_cntl/sleep/mod.rs

```diff
@@ -131,6 +131,71 @@ impl<'a, 'b> RtcioWakeupSource<'a, 'b> {
     }
 }
 
+macro_rules! uart_wakeup_impl {
+    ($num:literal) => {
+        paste::paste! {
+            #[doc = concat!("UART", $num, " wakeup source")]
+            ///
+            /// The chip can be woken up by reverting RXD for multiple cycles until the
+            /// number of rising edges is equal to or greater than the given value.
+            ///
+            /// Note that the character which triggers wakeup (and any characters before
+            /// it) will not be received by the UART after wakeup. This means that the
+            /// external device typically needs to send an extra character to trigger
+            /// wakeup before sending the data.
+            ///
+            /// After waking-up from UART, you should send some extra data through the UART
+            /// port in Active mode, so that the internal wakeup indication signal can be
+            /// cleared. Otherwise, the next UART wake-up would trigger with two less
+            /// rising edges than the configured threshold value.
+            ///
+            /// Wakeup from light sleep takes some time, so not every character sent to the
+            /// UART can be received by the application.
+            ///
+            /// This wakeup source can be used to wake up from light sleep only.
+            pub struct [< Uart $num WakeupSource >] {
+                threshold: u16,
+            }
+
+            impl [< Uart $num WakeupSource >] {
+                #[doc = concat!("Create a new instance of UART", $num, " wakeup source>") ]
+                ///
+                /// # Panics
+                ///
+                /// Panics if `threshold` is out of bounds.
+                pub fn new(threshold: u16) -> Self {
+                    if threshold > 1023 {
+                        panic!("Invalid threshold");
```

Comment:
I think in this case a panic gets much less into user's ways than having to check a `Result` which might just gets unwrapped